### PR TITLE
fix(replays): search is more robust now that we look inside of breadcrumb data.

### LINKED ISSUE #34805: Fix the replay details console search to look thru every aspect of the console log.
### Problem Statement

Search doesn't work quite right as certain parts of the message aren't found. Need to fix that so search works like a user would expect.

### Solution Brainstorm

_No response_

### PR DESCRIPTION
**Pull request by Danondso**
Closes #34805 

image
image
image

Created a more robust search function that accounts for all of what's shown within the console window.

### Legal Boilerplate

Look, I get it. The entity doing business as "Sentry" was incorporated in the State of Delaware in 2015 as Functional Software, Inc. and is gonna need some rights from me in order to utilize my contributions in this here PR. So here's the deal: I retain all rights, title and interest in and to my contributions, and by keeping this boilerplate intact I confirm that Sentry can use, modify, copy, and redistribute my contributions, under Sentry's choice of terms.

### CODE REVIEW

File: static/app/views/replays/detail/console/index.tsx [hunk 1]
```diff
@@ -30,11 +31,42 @@ function Console({breadcrumbs, startTimestamp = 0}: Props) {
...
+                default: {
+                  if (Array.isArray(value)) {
+                    doesMatch = value
+                      .join(' ')
+                      .toLowerCase()
+                      .includes(normalizedSearchTerm);
+                  }
+                }
+              }
+            });
```
Comment (Danondso) [inline static/app/views/replays/detail/console/index.tsx:64]: I wasn't sure of how else to manage searching thru record<string, any> so this is what I came up with. Would love to get feedback on a better way to handle it if anyone has any better ideas.

Comment (billyvg) [reply static/app/views/replays/detail/console/index.tsx:64]: can we extract this into a function and add a test for it as well?

Comment (Danondso) [reply static/app/views/replays/detail/console/index.tsx]: Yup that works, initially I had it like that but I neglected to check if log level had anything to filter on. So it wasn't working. All good now

File: tests/js/spec/views/replays/detail/console/index.spec.jsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+  it('filters console breadcrumbs by search term', () => {
+    const result = breadcrumbs.filter(breadcrumb =>
+      filterBreadcrumbs(breadcrumb, 'comp', [])
+    );
+
+    expect(result.length).toEqual(2);
+  });
+
+  it('filters console breadcrumbs by search term and log level', () => {
```
Comment (billyvg) [inline tests/js/spec/views/replays/detail/console/index.spec.jsx:97]: Could we add a test case where we search for an object's key or value inside of data.arguments?

File: static/app/views/replays/detail/console/index.tsx [hunk 3]
```diff
@@ -30,10 +41,8 @@ function Console({breadcrumbs, startTimestamp = 0}: Props) {
     () =>
       !searchTerm && logLevel.length === 0
         ? breadcrumbs
-        : breadcrumbs.filter(
-            breadcrumb =>
-              breadcrumb.message?.toLowerCase().includes(searchTerm) &&
-              logLevel.includes(breadcrumb.level)
+        : breadcrumbs.filter(breadcrumb =>
+            filterBreadcrumbs(breadcrumb, searchTerm, logLevel)
```
Comment (billyvg) [inline static/app/views/replays/detail/console/index.tsx:45]: I would probably change the function signature to be `filterBreadcrumbs(breadcrumbs: Crumb[], ...)` and include the logic from the ternary. That way we can add test cases for those conditions as well.

File: static/app/views/replays/detail/console/index.tsx [hunk 4]
```diff
@@ -21,6 +21,17 @@ interface Props {
 const getDistinctLogLevels = breadcrumbs =>
   Array.from(new Set<string>(breadcrumbs.map(breadcrumb => breadcrumb.level)));
 
+export const filterBreadcrumbs = (breadcrumb, searchTerm, logLevel) => {
```
Comment (billyvg) [inline static/app/views/replays/detail/console/index.tsx:24]: Let's add types here

File: static/app/views/replays/detail/console/index.tsx (hunk 3 above)
Comment (Danondso) [reply static/app/views/replays/detail/console/index.tsx]: Good call, adjusted.

File: static/app/views/replays/detail/console/index.tsx [hunk 5]
```diff
@@ -21,20 +21,33 @@ interface Props {
...
+export const filterBreadcrumbs = (
+  breadcrumbs: BreadcrumbTypeDefault[],
+  searchTerm: string,
+  logLevel: Array<string>
+) => {
+  if (!searchTerm && logLevel.length === 0) {
+    return breadcrumbs;
+  }
+  return breadcrumbs.filter(breadcrumb => {
+    const normalizedSearchTerm = searchTerm.toLowerCase();
```
Comment (ryan953) [inline static/app/views/replays/detail/console/index.tsx:33]:
```suggestion
  }
  const normalizedSearchTerm = searchTerm.toLowerCase();
  return breadcrumbs.filter(breadcrumb => {

```

**Review by billyvg — APPROVED** [review]

**Pull request merged**